woocommerce/google-listings-and-ads
# Automatically sync WooCommerce shipping settings with Merchant Center

### Changes proposed in this Pull Request:



This PR adds automatic syncing functionality for synchronizing the WooCommerce shipping zone settings to Google Merchant Center whenever there's a change applied by the user in their WooCommerce store settings.

There's a new async ActionSchedule job called `update_shipping_settings`. This job will run whenever we detect any changes to WooCommerce shipping settings and it will upload the updated settings to Google Merchant Center. Before scheduling and before processing (two times) it checks whether the MC account is connected AND the user has chosen the "automatic" syncing option for their shipping settings during onboarding.

To watch for WooCommerce settings changes, a new SyncerHooks class is introduced under the `Shipping` namespace. This class attaches an action to various WooCommerce and WordPress hooks and whenever it detects a change it schedules the `update_shipping_settings` job. WooCommerce uses various methods for updating the shipping settings (ajax, rest API, etc.); therefore, it was necessary to watch for multiple hooks. However, there is a flag in place to prevent the job from being scheduled more than once on each request.

This PR is based on the [main feature branch](https://github.com/woocommerce/google-listings-and-ads/pull/1494) for shipping integration milestone two.


### Detailed test instructions:


1. Make sure you have your MC account connected and finished the onboarding steps
2. Confirm that you have chosen the "Recommended" option for syncing your shipping settings automatically from WooCommerce
3. Go to WooCommerce settings and add a new shipping zone.
4. Add new methods to it. Edit then and add prices and save.
5. Add new locations and postcodes. Save the shipping zone.
6. Edit previously added methods or delete them. Save the shipping zone.
7. Edit the zone's locations or delete them and save the shipping zone. 
8. Add new shipping classes, assign prices to them in your shipping methods and then rename the classes or delete them.
9. Confirm that in all the test cases mentioned in steps 3-8, the `update_shipping_settings` job is scheduled in ActionScheduler.
10. Confirm in your MC account that the new shipping settings are reflected in your account's shipping section
11. Edit your free listings settings and change your shipping option to anything other than the "Recommended" option.
12. Repeat steps 3-8 and confirm that no update job is scheduled in ActionScheduler and your shipping settings remained unchained in MC.
13. Disconnect your MC account and check steps 3-8. No job should be scheduled to update the shipping settings.


### Additional details:


### Changelog entry

> Add - Automatically sync WooCommerce shipping settings with Merchant Center
